br/pkg/utils/suspend_importing.go

```diff
@@ -0,0 +1,136 @@
+}
+
+type SuspendImporting struct {
+	env  SuspendImportingEnv
+	name string
+}
+
+func NewSuspendImporting(name string, env SuspendImportingEnv) *SuspendImporting {
+	return &SuspendImporting{
+		env:  env,
+		name: name,
+	}
+}
+
+// DenyAllStores tries to deny all current stores' lightning execution for the period of time.
+// Returns a map mapping store ID to whether they are already denied to import tasks.
+func (d *SuspendImporting) DenyAllStores(ctx context.Context, dur time.Duration) (map[uint64]bool, error) {
+	return d.forEachStores(ctx, func() *import_sstpb.SuspendImportRPCRequest {
+		return &import_sstpb.SuspendImportRPCRequest{
+			ShouldSuspendImports: true,
+			DurationInSecs:       uint64(dur.Seconds()),
+			Caller:               d.name,
+		}
+	})
+}
+
+func (d *SuspendImporting) AllowAllStores(ctx context.Context) (map[uint64]bool, error) {
+	return d.forEachStores(ctx, func() *import_sstpb.SuspendImportRPCRequest {
+		return &import_sstpb.SuspendImportRPCRequest{
+			ShouldSuspendImports: false,
+			Caller:               d.name,
+		}
+	})
+}
+
+// forEachStores send the request to each stores reachable.
+// Returns a map mapping store ID to whether they are already denied to import tasks.
+func (d *SuspendImporting) forEachStores(ctx context.Context, makeReq func() *import_sstpb.SuspendImportRPCRequest) (map[uint64]bool, error) {
+	stores, err := d.env.GetAllStores(ctx)
+	if err != nil {
+		return nil, errors.Annotate(err, "failed to get all stores")
+	}
+
+	result := map[uint64]bool{}
+	for _, store := range stores {
+		cli, err := d.env.GetDenyLightningClient(ctx, store.Id)
+		if err != nil {
+			return nil, errors.Annotatef(err, "failed to get client for store %d", store.Id)
+		}
+		req := makeReq()
+		resp, err := cli.SuspendImportRPC(ctx, req)
+		if err != nil {
+			return nil, errors.Annotatef(err, "failed to deny lightning rpc for store %d", store.Id)
+		}
+		result[store.Id] = resp.AlreadySuspended
+	}
+	return result, nil
+}
+
+// HasKeptDenying checks whether a result returned by `DenyAllStores` is able to keep the consistency with last request.
+// i.e. Whether the store has some holes of pausing the import requests.
+func (d *SuspendImporting) ConsistentWithPrev(result map[uint64]bool) error {
+	for storeId, denied := range result {
+		if !denied {
+			return errors.Annotatef(berrors.ErrPossibleInconsistency, "failed to keep importing to store %d being denied, the state might be inconsistency", storeId)
+		}
+	}
+	return nil
+}
+
+func (d *SuspendImporting) Keeper(ctx context.Context, ttl time.Duration) error {
+	lastSuccess := time.Now()
+	t := time.NewTicker(ttl / DenyLightningUpdateFrequency)
+	defer t.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-t.C:
+			res, err := d.DenyAllStores(ctx, ttl)
+			if err != nil {
+				if time.Since(lastSuccess) < ttl {
```

Comment:
should it be time.Since(lastSuccess) < ttl * (DenyLightningUpdateFrequency - 1) / DenyLightningUpdateFrequency?

Reply:
I guess the next call will fail due to the check of consistency. 🤔️

Reply:
ok